File: 1460-number-of-substrings-containing-all-three-characters/1460-number-of-substrings-containing-all-three-characters.java
class Solution {
    public int numberOfSubstrings(String s) {
      
        int right = 0;
        int lastSeen[] = {-1,-1,-1};
        int count = 0;
        
        while(right < s.length()) {
            lastSeen[s.charAt(right) - 'a'] = right;
           
            if(lastSeen[0] != -1 && lastSeen[1] != -1 && lastSeen[2] != -1) {
                count += 1 + Math.min(Math.min(lastSeen[0],lastSeen[1]),lastSeen[2]);
            }
            right++;
        }

        return count;
    }
}

File: 1460-number-of-substrings-containing-all-three-characters/1460-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        left  = 0
        right = 0
        lastSeen = [0] * 3
        count = 0

        while(right < n):
            lastSeen[ord(s[right]) - ord('a')] += 1
            while(lastSeen[0] > 0 and lastSeen[1] > 0 and lastSeen[2] > 0):
                count += (n - right)
                lastSeen[ord(s[left]) - ord('a')] -= 1
                left += 1
            right += 1

        return count
        
        